### PR TITLE
new unstable option: -Zwrite-long-types-to-disk

### PR DESCRIPTION
**Pull request by mdibaiee**
This option guards the logic of writing long type names in files and instead using short forms in error messages in rustc_middle/ty/error behind a flag. The main motivation for this change is to disable this behaviour when running ui tests.

This logic can be triggered by running tests in a directory that has a long enough path, e.g. /my/very-long-path/where/rust-codebase/exists/

This means ui tests can fail depending on how long the path to their file is.

Some ui tests actually rely on this behaviour for their assertions, so for those we enable the flag manually.

### CODE REVIEW

**Comment by pnkfelix** [comment]
It looks like you missed updating the stderr for  tests/ui/recursion/issue-83150.rs 

I plan to help this PR move along by contributing patches to it (I tend to have long paths to my working directories, so landing this will help unblock my own development)

File: compiler/rustc_middle/src/ty/error.rs [hunk 1]
```diff
@@ -309,11 +309,12 @@ impl<'tcx> Ty<'tcx> {
 
 impl<'tcx> TyCtxt<'tcx> {
     pub fn ty_string_with_limit(self, ty: Ty<'tcx>, length_limit: usize) -> String {
-        let mut type_limit = 50;
         let regular = FmtPrinter::new(self, hir::def::Namespace::TypeNS)
             .pretty_print_type(ty)
             .expect("could not write to `String`")
             .into_buffer();
+
+        let mut type_limit = 50;
```
Comment (pnkfelix) [inline compiler/rustc_middle/src/ty/error.rs:317]: I don't think this change is paying for itself. (I.e. there's a cost with changing code unnecessarily, in terms of impeding people who e.g. use `git annotate` as part of their workflow.) So I'd revert this back to the prior form.

**Comment by mdibaiee** [comment]
I can't figure out what the root cause of this last error I am getting is (to be fair I didn't put too much time investigating). The option I have added is `untracked`, but the error seems to be related to tracked options. 🤨 If someone can explain I can see about resolving it

File: compiler/rustc_interface/src/tests.rs [hunk 2]
```diff
@@ -738,6 +738,7 @@ fn test_unstable_options_tracking_hash() {
     untracked!(unstable_options, true);
     untracked!(validate_mir, true);
     untracked!(verbose, true);
+    untracked!(write_long_types_to_disk, true);
```
Comment (compiler-errors) [inline compiler/rustc_interface/src/tests.rs:741]: This needs to be set to a value that's *not* default

```suggestion
    untracked!(write_long_types_to_disk, false);
```

**Review by compiler-errors — CHANGES_REQUESTED** [review]
This looks good from the compiler side, just had one concern about UI tests:

Most of these tests don't seem to be testing long-type-names. Can you remove `-Zwrite-long-types-to-disk=yes` from all of them except for like `long-E0308.rs`, and all the other normalization stuff like

```
// normalize-stderr-test: "long-type-\d+" -> "long-type-hash"
```

and bless all the tests?

**Comment by mdibaiee** [comment]
@compiler-errors done! I initially added `write-long-types-to-disk=yes` to some other tests because I figured they always will produce long-enough types that would trigger the logic. I have now reverted that change and now only the test for this specific logic uses this flag.

**Review by compiler-errors — APPROVED** [review]

**Pull request merged**